Fixes #28193 - Handle pulp3 services

johnpmitsch opened the pull request:
This adds the ability to handle pulp3 services. A service is used to determine if pulp3 is present as it currently does not run off rpm's in our environment. This could change to be based off an RPM later on when pulp3 is packaged and included with Katello. This also adds the ability to handle glob operators for services and optionally use the `--all` flag for systemctl

johnpmitsch commented on definitions/features/pulp3.rb:
```diff
@@ -0,0 +1,22 @@
+      system_service('pulpcore-resource-manager', 10),
+      system_service('pulpcore-worker@*', 20, :all => true),
+      system_service('redis', 30),
+      system_service('httpd', 30)
```
@jlsherrill what do you think about the order here? Do we want to start and stop services in a particular order?

This is the order with the current values:

```
| displaying rh-mongodb34-mongod                   [OK]                         
| displaying postgresql                            [OK]                         
| displaying pulpcore-api                          [OK]                         
/ displaying pulpcore-content                      [OK]                         
/ displaying pulpcore-resource-manager             [OK]                         
/ displaying qdrouterd                             [OK]                         
/ displaying qpidd                                 [OK]                         
/ displaying squid                                 [OK]                         
/ displaying pulp_celerybeat                       [OK]                         
/ displaying pulp_resource_manager                 [OK]                         
/ displaying pulp_streamer                         [OK]                         
/ displaying pulp_workers                          [OK]                         
- displaying pulpcore-worker@*                     [OK]                         
- displaying tomcat                                [OK]                         
- displaying httpd                                 [OK]                         
- displaying puppetserver                          [OK]                         
- displaying redis                                 [OK]                         
- displaying foreman-proxy                         [OK]        
```

jlsherrill replied:
We might ask the pulp team if it matters, i might could see it being better to shut down the api, then the resource-manager and then the workers, but in practice it may not matter.   I'd reach out to the pulp team and ask if they have a preference.

jlsherrill replied:
also, does the order reverse itself when starting vs stopping?

mbacovsky replied:
It does reverse

mbacovsky commented:
@johnpmitsch  thanks for the patch! Looks good.

There is one thing that we should consider. There are various features/procedures/checks that are conditioned by presence of pulp. We use `feature(:pulp)` to test the pulp presence. If there is going to be new feature for Pulp 3 we should revisit the code.

For Mongo 2 vs Mongo 3 we chose to have both in a single `feature(mongo)` but for Pulp it may be different and separating the two features may be useful. Would it make sense to have `feature(:pulp2)` and `feature(:pulp3)` and perhaps `feature(:instance).pulp?` defined as `feature(:pulp3) || feature(:pulp2)`?

Any other ideas?
@kgaikwad, @upadhyeammit?

johnpmitsch commented:
@mbacovsky  Thanks, that is a really good point and we will have to consider this as we add more pulp3 functionality to foreman-maintain such a backups, managing upgrades, and perhaps parts of the pulp 2->3 migration.

> Would it make sense to have `feature(:pulp2)` and `feature(:pulp3)` and perhaps `feature(:instance).pulp`

This sounds like a good idea, +1 to keeping them separate features as they are two very different apps and I think it would be a lot of `if X else Y` to manage them in one feature. Though we could have a single indicator if any pulp instance is present as you suggested. 

Let me see what I can do!

jlsherrill commented:
One thing to ask is, when would we want to know the some version of pulp is
enabled but not actually care which version?  I haven't had a chance to
look through all of the foreman maintain code yet, but are there many
examples of that?

On Mon, Nov 18, 2019, 9:49 AM John Mitsch <notifications@github.com> wrote:

> @mbacovsky <https://github.com/mbacovsky> Thanks, that is a really good
> point and we will have to consider this as we add more pulp3 functionality
> to foreman-maintain such a backups, managing upgrades, and perhaps parts of
> the pulp 2->3 migration.
>
> Would it make sense to have feature(:pulp2) and feature(:pulp3) and
> perhaps feature(:instance).pulp
>
> This sounds like a good idea, +1 to keeping them separate features as they
> are two very different apps and I think it would be a lot of if X else Y
> to manage them in one feature. Though we could have a single indicator if
> any pulp instance is present as you suggested.
>
> Let me see what I can do!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman_maintain/pull/290?email_source=notifications&email_token=AADAORIM56DS2QQ7YBTTLGTQUKTOLA5CNFSM4JN6OCKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEKVTQI#issuecomment-555047361>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADAORNDBLWFYT5X5IETPIDQUKTOLANCNFSM4JN6OCKA>
> .
>

johnpmitsch commented:
@jlsherrill sounds like the database should be started first:
```
dkliban | jomitsch: the database should be started first. the rest of the services can be brought up after that in 
         | any order                                                                                                
 dkliban | database and redis should be started before any other services are started                               
```

jlsherrill commented:
yeah, and it will in the end share a database server with foreman and candlepin, so i guess it can be handled in the same way

jlsherrill commented:
we may need to either hold off on supporting the pulp3 database for now, or wait to merge this until the postgresql 10/11 work is done.

kgaikwad commented:
> For Mongo 2 vs Mongo 3 we chose to have both in a single `feature(mongo)` but for Pulp it may be different and separating the two features may be useful. Would it make sense to have `feature(:pulp2)` and `feature(:pulp3)` and perhaps `feature(:instance).pulp?` defined as `feature(:pulp3) || feature(:pulp2)`?
> Any other ideas?
> @kgaikwad, @upadhyeammit?

@mbacovsky, sounds good to me +1

johnpmitsch commented:
Updated to change the pulp feature to pulp2 and use a `feature(:instance).pulp` for checking both. I've updated the codebase for this change. For the backup, I'm just assuming we are backing up pulp2, the pulp3 backup support is a different task on our list

> we may need to either hold off on supporting the pulp3 database for now, or wait to merge this until the postgresql 10/11 work is done.

@jlsherrill What will this look like, will it be another service? I think this could be a separate task as it feels like there will be another feature added for postgres 10 here, separate from the pulp feature.

jlsherrill commented:
Yes, i think we wait till postgresql 10 support lands, there will only be 1 instance that pulp/foreman/candlepin share, so i imagine it will look similar to: https://github.com/theforeman/foreman_maintain/blob/master/definitions/features/candlepin_database.rb

jlsherrill commented:
and yes, i agree pulp3 backup is a different task.  We may need to wait till we can actually provision a box with pulp3 to start working on it

jlsherrill commented:
@kgaikwad @mbacovsky any more thoughts?

johnpmitsch commented on definitions/procedures/service/base.rb:
```diff
@@ -15,7 +15,7 @@ def self.common_params(context)
       def run_service_action(action, options)
         action_noun = feature(:service).action_noun(action).capitalize
-        puts "#{action_noun} the following service(s):\n"
+        puts "\n#{action_noun} the following service(s):"
```
This spacing has bothered me for a while, I can do a separate PR to change, but its just a small thing so I included here! :cat:

johnpmitsch commented:
@kgaikwad @mbacovsky  I don't mean to be a squeaky wheel, but any chance we can look at this this week?

jlsherrill commented on definitions/features/pulp3.rb:
```diff
@@ -0,0 +1,22 @@
+      system_service('pulpcore-api', 10),
+      system_service('pulpcore-content', 10),
+      system_service('pulpcore-resource-manager', 10),
+      system_service('pulpcore-worker@*', 20, :all => true),
```
I tried a  'foreman-maintain service restart'  and it didn't seem to properly start the pulpcore-worker services, did it work for you?

ianballou replied:
I see this issue as well

ianballou replied:
looks like the `:all => true` isn't making it to the actual `systemctl` command.

johnpmitsch commented:
@ianballou @jlsherrill nice catch, I updated in a separate commit. I guess I was checking the pulp2 workers in my testing! The behavior makes sense because you need --all for using globs with `systemctl start` and do not need it for `systemctl stop`, so they were stopped but not started, since the globs weren't registering with `start` because `--all` wasn't being added to the command. (I'm not sure the reason for why systemctl works this way)

I'm not sure if there is any concern of having both `@options` and `options`, but using the options hash for the instance works well.

ianballou commented:
Latest commit fixed the Pulp 3 worker issue I was having :+1:

jlsherrill commented on lib/foreman_maintain/utils/service/systemd.rb:
```diff
@@ -1,17 +1,20 @@
-          "systemctl #{action} #{@name}"
+          cmd = "systemctl #{action} #{@name}"
+          cmd += ' --all' if all
+          cmd
```
hrmm, i just realized this means that --all will be passed to enable/disable which i'm not sure makes sense

in fact i realized that foreman-maintain even provides enable/disable support....  
which i don't think will do the right thing:

```
systemctl enable 'pulpcore-worker@*'  --all

Created symlink from /etc/systemd/system/multi-user.target.wants/pulpcore-worker@\x2a.service to /usr/lib/systemd/system/pulpcore-worker@.service.
```

so i think it might make more sense if the start/stop commands pass :all in through the options param to command.  And then we need to figure out what to do with enable/disable.  I'm not sure how its supposed to know which pulpworkers to enable if they've been disabled.

jlsherrill replied:
the simplest thing is to not allow disabling/enabling of pulpcore-workers, but that seems kinda like a cop out.  I'm not entirely sure why this functionality is provided.

jlsherrill commented on lib/foreman_maintain/utils/service/systemd.rb:
```diff
@@ -1,17 +1,25 @@
       end
 
       def command(action, options = {})
+        if %w[enable disable].include?(action) && @name.include?('pulpcore-worker@*')
```
I think it'd be better to pass this in as an option, :disable_enablement_control => true  or something like that, and have it be controlled here.  I'd probably handle that in the enable and disable methods rather than in this command method as well.

I'd think that this command module should not have any service-specific knowledge.

johnpmitsch replied:
thats a good point, let me look into that

johnpmitsch replied:
@jlsherrill updated to take an option and use that to skip

johnpmitsch replied:
err... hold on :)

johnpmitsch replied:
@jlsherrill ok, its now updated to only skip on enable and disable with that option :)

johnpmitsch commented:
@jlsherrill  updated to skip enable/disable for pulpcore_workers and display a message.

@mbacovsky @kgaikwad the way the pulp3 workers are set up currently is that they are enabled in systemd through the installer and there is a variable amount of them. So by disabling them, they "disappear" and have to enabled one by one, for example `pulpcore_worker@1` and `pulpcore_worker@2`. Using `pulpcore_worker@*` won't enable the correct amount of workers with the correct names, it will just literally register the `*`.

  We don't know the number of workers (its not in a config somewhere like pulp2) so we couldn't come up with a way to handle this, so we are just skipping the enable/disable commands for them and displaying a warning message. I don't think its a huge deal since enable/disable aren't probably heavily used commands, and start, stop, and status work fine. I just want to make sure you all are aware of the background of why it's handled like this in the code. Hopefully we can improve this on our end in the installer, giving a better way to manage these services (like the pulp_workers service in pulp2) and update foreman-maintain with that way.

johnpmitsch commented:
If we do wind up storing the number of workers somewhere, we can enable `pulpcore_worker@{1..N}`, where `N` is the number of workers, this syntax works with systemctl

(more commenting this as future reference if we come back to this later)

jlsherrill commented:
This looks good, but i will have to test at a certain time.  @ianballou if you want to re-test that'd be very helpful!
So ack from a code perspective from me.  @kgaikwad you might want to take a 2nd glance at this as well

jlsherrill commented:
@johnpmitsch also the test failure seems unrelated, can you restart it?

johnpmitsch commented:
> This looks good, but i will have to test at a certain time. @ianballou if you want to re-test that'd be very helpful!
> So ack from a code perspective from me. @kgaikwad you might want to take a 2nd glance at this as well

Sounds good


> @johnpmitsch also the test failure seems unrelated, can you restart it?

They seem to be consistently failing with an possible build issue, maybe @kgaikwad has an idea of if this is a known issue?

kgaikwad commented:
@johnpmitsch,
Due to `mocha` gem version, now test helper will need to update `require "mocha/mini_test" ` to `"mocha/minitest"`.

@upadhyeammit is going to create a separate PR for this.

ianballou commented:
I've tested this and noticed that my `pulp-worker@0` is inactive after the running service restart.  Workers 1 and 2 and the rest of the services are up and running though.  If this is expected then it should be all good!

jlsherrill commented:
@ianballou was the 0 worker actually enabled?  
``` 
systemctl list-units -t service --full  | grep pulpcore-worker
```
the pulp installer only enables workers 1 & 2, so i wouldn't expect 0 to be started by this script unless you had enabled it manually prior.

ianballou commented:
@jlsherrill ah ok that was it! I'm not sure why worker 0 was even running to be honest.  ACK

kgaikwad commented:
@johnpmitsch,
Could you rebase this branch as I have merged #297?
